python/descriptor_containers.c

```diff
@@ -0,0 +1,670 @@
+};
+
+static PyType_Spec PyUpb_DescriptorIterator_Spec = {
+  PYUPB_MODULE_NAME ".DescriptorIterator",   // tp_name
+  sizeof(PyUpb_DescriptorIterator),          // tp_basicsize
+  0,                                         // tp_itemsize
+  Py_TPFLAGS_DEFAULT,                        // tp_flags
+  PyUpb_DescriptorIterator_Slots,
+};
+
+// -----------------------------------------------------------------------------
+// GenericSequence
+// -----------------------------------------------------------------------------
+
+typedef struct {
+  PyObject_HEAD
+  const PyUpb_GenericSequence_Funcs *funcs;
+  const void *parent;    // upb_msgdef*, upb_symtab*, etc.
+  PyObject *parent_obj;  // Python object that keeps parent alive, we own a ref.
+} PyUpb_GenericSequence;
+
+PyObject *PyUpb_GenericSequence_New(
+    const PyUpb_GenericSequence_Funcs *funcs, const void *parent,
+    PyObject *parent_obj) {
+  PyUpb_ModuleState *s = PyUpb_ModuleState_Get();
+  PyUpb_GenericSequence *seq =
+      (PyUpb_GenericSequence *)PyType_GenericAlloc(s->generic_sequence_type, 0);
+  seq->funcs = funcs;
+  seq->parent = parent;
+  seq->parent_obj = parent_obj;
+  Py_INCREF(parent_obj);
+  return &seq->ob_base;
+}
+
+PyUpb_GenericSequence *PyUpb_GenericSequence_Get(PyObject *obj) {
+  assert(Py_TYPE(obj) == PyUpb_ModuleState_Get()->generic_sequence_type);
+  return (PyUpb_GenericSequence*)obj;
+}
+
+static void PyUpb_GenericSequence_Dealloc(PyObject *_self) {
+  PyUpb_GenericSequence *self = PyUpb_GenericSequence_Get(_self);
+  Py_CLEAR(self->parent_obj);
+  PyUpb_Dealloc(self);
+}
+
+static Py_ssize_t PyUpb_GenericSequence_Length(PyObject* _self) {
+  PyUpb_GenericSequence *self = PyUpb_GenericSequence_Get(_self);
+  return self->funcs->get_elem_count(self->parent);
+}
+
+static PyObject *PyUpb_GenericSequence_GetItem(PyObject *_self,
+                                               Py_ssize_t index) {
+  PyUpb_GenericSequence *self = PyUpb_GenericSequence_Get(_self);
+  Py_ssize_t size = self->funcs->get_elem_count(self->parent);
+  if (index < 0 || index >= size) {
+    PyErr_Format(PyExc_IndexError, "list index (%zd) out of range", index);
+    return NULL;
+  }
+  const void *elem = self->funcs->index(self->parent, index);
+  return self->funcs->get_elem_wrapper(elem);
+}
+
+// A sequence container can only be equal to another sequence container, or (for
+// backward compatibility) to a list containing the same items.
+// Returns 1 if equal, 0 if unequal, -1 on error.
+static int PyUpb_GenericSequence_IsEqual(PyUpb_GenericSequence *self,
+                                         PyObject *other) {
+  // Check the identity of C++ pointers.
+  if (PyObject_TypeCheck(other, Py_TYPE(self))) {
+    PyUpb_GenericSequence *other_seq = (void *)other;
+    return self->parent == other_seq->parent && self->funcs == other_seq->funcs;
+  }
+
+  if (!PyList_Check(other)) return 0;
+
+  // return list(self) == other
+  int n = PyUpb_GenericSequence_Length((PyObject*)self);
```

Comment:
Python lists are not restricted, but `PyUpb_GenericSequence` objects are (everything in the vtable uses `int` for index, following the conventions of `upb/def.h`.

Reply:
but other might not be a upb list and so you might be truncating here

Reply:
oh, wait, you are avoiding capturing other as an int.  This is a bit more subtle than desired though, please add a comment (or use size_t since you know it won't be negative)